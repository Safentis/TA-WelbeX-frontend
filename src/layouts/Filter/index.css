.filter {
  align-items: center;
  display: flex;
}

.filter_active {
  background-color: var(--main-grey-color);
  transition: .3s;
}

.filter__item {
  align-items: center;
  display: grid;
  justify-content: center;
  transition: 0.3s;
}

.filter__submenu {
  align-items: stretch;
  grid-template-columns: 1.5fr 1.5fr 1.5fr;
  justify-items: center;
  padding: 0 .5rem;
  width: 100%;
}

.filter__item_noactive {
  opacity: 0;
  transform: translateX(-10rem);
  z-index: -10;
}

.filter__item_active {
  opacity: 1;
  transform: translateX(0rem);
  z-index: 0;
}

.filter__nested {
  display: flex;
  align-items: center;
}