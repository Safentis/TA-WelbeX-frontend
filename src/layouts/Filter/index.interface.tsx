import { Filter } from '../../application/App/App.interface';

export interface Props {
  filter: Filter;
  setFilter: Function;
}
