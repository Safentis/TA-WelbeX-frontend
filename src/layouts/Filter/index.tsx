import React, { FC, useState } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faPuzzlePiece,
  faAngleRight,
  faAngleLeft,
  faEquals,
  faFilter,
  faTimes,
} from '@fortawesome/free-solid-svg-icons';

import Button from '../../components/Button/Button';
import FilterRadio from './FilterRadio/FilterRadio';
import FilterSearch from './FilterSearch/FilterSearch';

import './index.css';
import { Props } from './index.interface';

/**
 * Filter components that contains all options of
 * sort, you may choose filter by 3 stage
 * 1) column
 * 2) signals (more, lest, equal and keep)
 * 3) text field
 *
 * @param {FilterInterface} filter
 * @param {Function} setFilter
 * @returns {React.ReactElement} - returns react element
 */

const Filter: FC<Props> = ({
  filter,
  setFilter,
}): React.ReactElement => {
  //* ----------------------------------------------
  //* Show or hide sub menu with filter
  const [isOpen, setIsOpen] = useState<boolean>(false);
  const handleClick = (): void => {
    setIsOpen(!isOpen);
  };

  //* Dynamic styles
  const filterItemActiveCls: string = isOpen
    ? 'filter__item_active'
    : 'filter__item_noactive';
  const btnActiveCls: string = isOpen ? 'button_active' : '';
  const filterActiveCls: string = isOpen ? 'filter_active' : '';

  //* ----------------------------------------------------
  const { column, operation, search } = filter;
  const operationIcons = [
    faAngleRight,
    faAngleLeft,
    faEquals,
    faPuzzlePiece,
  ];

  //* In this case, we are handling radio buttons
  const handleChange = (event: React.ChangeEvent): void => {
    //* We are creating copy of state;
    const newState = { ...filter };
    const targer = event.target as HTMLInputElement;
    const chooseValue: string = targer.value;
    const chooseName: string = targer.name;

    //* We are choosing category about column or operation
    const filterObject =
      chooseName === 'column' ? newState.column : newState.operation;

    //* If it's value we are seting true
    //* else false
    for (let field in filterObject) {
      field !== chooseValue
        ? (filterObject[field] = false)
        : (filterObject[chooseValue] = true);
    }

    //* And finaly, we are seting new state
    setFilter(newState);
  };

  const handleSearch = (event: React.ChangeEvent): void => {
    const newState = { ...filter };
    const targer = event.target as HTMLInputElement;

    //* We are changing search value
    newState.search = targer.value;
    setFilter(newState);
  };

  return (
    <ul className={`filter ${filterActiveCls}`}>
      <li className="filter__item">
        <Button className={btnActiveCls} handleClick={handleClick}>
          <FontAwesomeIcon icon={isOpen ? faTimes : faFilter} />
        </Button>
      </li>
      <li
        className={`filter__item filter__submenu ${filterItemActiveCls}`}
      >
        <FilterRadio
          className="filter__nested"
          handleChange={handleChange}
          fields={column}
          name="column"
        />
        <FilterRadio
          className="filter__nested"
          handleChange={handleChange}
          icons={operationIcons}
          fields={operation}
          name="operation"
        />
        <FilterSearch handleSearch={handleSearch} value={search} />
      </li>
    </ul>
  );
};

export default Filter;
