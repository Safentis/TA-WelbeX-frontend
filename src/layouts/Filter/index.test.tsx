import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';

import Filter from './index';

import { Props } from './index.interface';

const setUp = (props: Props) => {
  return shallow(<Filter {...props} />);
};

describe('<Filter />', () => {
  let component: ShallowWrapper<
    any,
    Readonly<{}>,
    React.Component<{}, {}, any>
  >;
  let props: Props = {
    filter: {
      column: {
        name: false,
        number: false,
        distance: false,
      },
      operation: {
        more: false,
        less: false,
        equal: false,
        contains: false,
      },
      search: '',
    },
    setFilter: () => {},
  };

  beforeEach(() => {
    component = setUp(props);
  });

  it('component does match to the snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  describe('Markup: ', () => {
    it('has 1 ul element', () => {
      expect(component.find('ul')).toBeTruthy();
    });

    it('has 1 li element', () => {
      expect(component.find('li')).toBeTruthy();
    });

    it('has 1 Button element', () => {
      expect(component.find('Button')).toBeTruthy();
    });

    it('has 1 FontAwesomeIcon element', () => {
      expect(component.find('FontAwesomeIcon')).toBeTruthy();
    });

    it('has 2 FilterRadio element', () => {
      expect(component.find('FilterRadio').length === 2).toBeTruthy();
    });

    it('has 1 FilterSearch element', () => {
      expect(component.find('FilterSearch')).toBeTruthy();
    });
  });
});
