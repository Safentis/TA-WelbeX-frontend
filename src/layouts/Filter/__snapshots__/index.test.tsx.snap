// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Filter /> component does match to the snapshot 1`] = `
<ul
  className="filter "
>
  <li
    className="filter__item"
  >
    <Button
      className=""
      handleClick={[Function]}
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              512,
              512,
              Array [],
              "f0b0",
              "M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z",
            ],
            "iconName": "filter",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
    </Button>
  </li>
  <li
    className="filter__item filter__submenu filter__item_noactive"
  >
    <FilterRadio
      className="filter__nested"
      fields={
        Object {
          "distance": false,
          "name": false,
          "number": false,
        }
      }
      handleChange={[Function]}
      name="column"
    />
    <FilterRadio
      className="filter__nested"
      fields={
        Object {
          "contains": false,
          "equal": false,
          "less": false,
          "more": false,
        }
      }
      handleChange={[Function]}
      icons={
        Array [
          Object {
            "icon": Array [
              256,
              512,
              Array [],
              "f105",
              "M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z",
            ],
            "iconName": "angle-right",
            "prefix": "fas",
          },
          Object {
            "icon": Array [
              256,
              512,
              Array [],
              "f104",
              "M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z",
            ],
            "iconName": "angle-left",
            "prefix": "fas",
          },
          Object {
            "icon": Array [
              448,
              512,
              Array [],
              "f52c",
              "M416 304H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32zm0-192H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z",
            ],
            "iconName": "equals",
            "prefix": "fas",
          },
          Object {
            "icon": Array [
              576,
              512,
              Array [],
              "f12e",
              "M519.442 288.651c-41.519 0-59.5 31.593-82.058 31.593C377.409 320.244 432 144 432 144s-196.288 80-196.288-3.297c0-35.827 36.288-46.25 36.288-85.985C272 19.216 243.885 0 210.539 0c-34.654 0-66.366 18.891-66.366 56.346 0 41.364 31.711 59.277 31.711 81.75C175.885 207.719 0 166.758 0 166.758v333.237s178.635 41.047 178.635-28.662c0-22.473-40-40.107-40-81.471 0-37.456 29.25-56.346 63.577-56.346 33.673 0 61.788 19.216 61.788 54.717 0 39.735-36.288 50.158-36.288 85.985 0 60.803 129.675 25.73 181.23 25.73 0 0-34.725-120.101 25.827-120.101 35.962 0 46.423 36.152 86.308 36.152C556.712 416 576 387.99 576 354.443c0-34.199-18.962-65.792-56.558-65.792z",
            ],
            "iconName": "puzzle-piece",
            "prefix": "fas",
          },
        ]
      }
      name="operation"
    />
    <FilterSearch
      handleSearch={[Function]}
      value=""
    />
  </li>
</ul>
`;
